src/components/Filter/Filter.js

```diff
@@ -1,113 +1,116 @@
 import './Filter.css';
 
-import React, { Component } from 'react';
+import React, { useReducer } from 'react';
 import classNames from 'classnames';
 import AutoComplete from '../AutoComplete/AutoComplete';
 import Input from '../Input/Input';
 import Switch from '../Switch/Switch';
 
 import { generateLists } from '../../listsGenerator';
 
-export default class Filter extends Component {
-  state = {
-    tag: '',
-    country: '',
-    name: '',
-    language: '',
-    showFilters: false,
-  };
+const Filter = props => {
+  const [state, setState] = useReducer(
+    (state, newState) => ({ ...state, ...newState }),
+    {
+      tag: '',
+      country: '',
+      name: '',
+      language: '',
+      showFilters: false,
+    }
+  );
 
-  onTagSelect = tag => {
-    this.setState({ tag });
-    this.props.onTagSelected(tag);
+  const onTagSelect = tag => {
+    setState({ tag });
+    props.onTagSelected(tag);
   };
 
-  onCountrySelect = country => {
-    this.setState({ country });
-    this.props.onCountrySelected(country);
+  const onCountrySelect = country => {
```

Comment:
We should use `useCallback` for all these functions, by using `useCallback` the components will receive the same instance on every render, preventing from rendering many times

Reply:
@crysfel it's not overly obvious, but setState is actually the callback from `useReducer` which invokes `useCallback` under the hood.

Reply:
@crysfel it's not overly obvious, but setState is actually the callback from `useReducer` which invokes `useCallback` under the hood.

Reply:
Not sure if I'm following you 😅 (It's very possible I didn't explain myself well) What I asked was about doing something like this:

```suggestion
const onCountrySelect = useCallback(country => {
  setState({ country });
  props.onCountrySelected(country);
}, [country])
```

This will memoize this function as we will get the same on every render and a new one when `country` gets updated, right?. 

Here's an example: *When the user types into the technology input the state is going to change, this will trigger a re-render, right? If we don't use `useCallback`, a new `onCountrySelect` function will be created, therefore the country input will render again (because `onCountrySelect` is different), but we don't need to re-render that field again, given that the value remains the same, right? We only need to render the technology field in this scenario*

Please correct me if I'm wrong, I might be missing something probably 😂

Reply:
@crysfel @brentmclark can we do a short chat to "close the deal" and complete that PR? Thanks!

Reply:
@crysfel thanks for the detailed explanation.  I can see now that my interpretation of the issue was way off.  The dispatch function returned from `useReducer` which we call `setState` in this implementation, is guaranteed to retain its function identity, *but* the outer function does not.  So, the scenario you've described is definitely happening and it would be wise to guard against it using `useCallback` as you've suggested.

In other news, I am going to stop reviewing PRs on my phone...